### PR TITLE
add derivative operator for Heaviside space

### PR DESCRIPTION
**Pull request by ngpooifung**

### CODE REVIEW

File: src/Spaces/Singularities/HeavisideSpace.jl [hunk 1]
```diff
@@ -112,7 +112,24 @@ function differentiate(f::Fun{<:HeavisideSpace})
...
+function Derivative(H::HeavisideSpace)
+    ConcreteDerivative(H)
+end
+
+bandinds(D::ConcreteDerivative{H}) where {H<:HeavisideSpace} = (-1,0)
+rangespace(D::ConcreteDerivative{H}) where {H<:HeavisideSpace} = domain(D).points[end]==Inf?DiracSpace(domain(D).points[1:end-1]):DiracSpace(domain(D).points)
 
+function getindex(D::ConcreteDerivative{H,T},k::Integer,j::Integer) where {H<:HeavisideSpace,T}
+    if k==j
+        1
```
Comment (dlfivefifty) [inline src/Spaces/Singularities/HeavisideSpace.jl]: Use one(T) to ensure it works with other types

**Comment by ngpooifung** [comment]
this can now be merged, is there anything I should look into next

**Comment by dlfivefifty** [comment]
Look at the following:
1. Solving simple ODEs. For example, the following should work: `S = Heavisidespace([-1.,0.,1.]); A = [ldirichlet(S); Derivative(S)]; A \ [0.,δ]`
2. Look at derivatives for `SplineSpace{1}`. This is the space of piecewise affine functions (trapezoidal approximation), and its derivative should be a `HeavisideSpace`.
3. Start looking at `ContinuousSpace`, which is the space of piecewise continuous polynomials. For ODEs like `u + u' = δ` it is not enough to have just `HeavisideSpace`, we need higher degree polynomials. I think a simple way of doing this is to use a `SumSpace` of a `HeavisideSpace` with a `ContinuousSpace`.

**Comment by dlfivefifty** [comment]
This PR added two warnings when calling `using ApproxFun`. Please fix these warnings and make sure you test the code next time.

**Comment by ngpooifung** [comment]
the first one doesn’t work somehow, it returns[cid:CBC1FBE8-D3DD-48D4-9864-66F3E0B9D3D5]

在 2018年8月2日，09:02，Sheehan Olver <notifications@github.com<mailto:notifications@github.com>> 写道：


Look at the following:

  1.  Solving simple ODEs. For example, the following should work: S = Heavisidespace([-1.,0.,1.]); A = [ldirichlet(S); Derivative(S)]; A \ [0.,δ]
  2.  Look at derivatives for SplineSpace{1}. This is the space of piecewise affine functions (trapezoidal approximation), and its derivative should be a HeavisideSpace.
  3.  Start looking at ContinuousSpace, which is the space of piecewise continuous polynomials. For ODEs like u + u' = δ it is not enough to have just HeavisideSpace, we need higher degree polynomials. I think a simple way of doing this is to use a SumSpace of a HeavisideSpace with a ContinuousSpace.

**Pull request merged**